soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -205,12 +218,157 @@ static int dt_init(const struct device *dev)
+ * @return false Returned when device is still busy when expected timeline elapsed.
+ */
+static bool wait_ipc_device_not_busy(uint32_t usec_to_wait, uint32_t retries)
+{
+	if (pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV)) {
+		while (retries > 0) {
+			k_busy_wait(usec_to_wait);
+			if (!pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV))
+				return true;
+			retries--;
+		}
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks. Returning error code from functions
+ * invoked during power stansition, does not release the pm_device_state_lock.
+ * It must be released manually by the caller during error handling process.
+ *
+ * @note Power state of the IPC Device is changed atomically.
+ * PM lock is taken at the start of each power transition to prevent power state change
+ * if the process fails.
+ * If IPC Device performs hardware operation, this function waits for 5 seconds in total
+ * for IPC Device to become non-busy (to finish all pending transactions).
+ * Note: when devices change power state using pm_suspend_devices() function, busy devices are
+ * ignored. However, calling pm_device_action_run() directly on the device, does not prevent
+ * busy device from power transition.
+ */
+static int ipc_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct ipc_control_driver_api *api =
+		(const struct ipc_control_driver_api *)dev->api;
+
+	/* wait up to 5 seconds for IPC device to become non-busy */
+	if (!wait_ipc_device_not_busy(1000000, 5))
+		return -EBUSY;
+
+	int ret = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_state_lock(dev);
+		if (api->suspend_fn) {
+			ret = api->suspend_fn(dev, api->suspend_fn_args);
+			if (ret)
+				break;
+		}
+		pm_device_state_unlock(dev);
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		pm_device_state_lock(dev);
+		irq_enable(DT_IRQN(INTEL_ADSP_IPC_HOST_DTNODE));
+		if (!irq_is_enabled(DT_IRQN(INTEL_ADSP_IPC_HOST_DTNODE))) {
+			ret = -ECANCELED;
+			break;
+		}
+		ace_ipc_intc_unmask();
+		ret = intel_adsp_ipc_init(dev);
+		if (ret)
+			break;
+		if (api->resume_fn) {
+			ret = api->resume_fn(dev, api->resume_fn_args);
+			if (ret)
+				break;
+		}
+		pm_device_state_unlock(dev);
+		break;
+	case PM_DEVICE_ACTION_TURN_OFF:
+		break;
+	case PM_DEVICE_ACTION_TURN_ON:
+		break;
+	default:
+		return ret;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Callback functions to be executed by Zephyr application
+ * during IPC device suspend and resume.
+ */
+static const struct ipc_control_driver_api ipc_power_control_api = {
+	.resume_fn = NULL,
+	.resume_fn_args = NULL,
+	.suspend_fn = NULL,
+	.suspend_fn_args = NULL
+};
+
+PM_DEVICE_DT_DEFINE(INTEL_ADSP_IPC_HOST_DTNODE, ipc_pm_action);
```

Comment:
This could be out of this `#ifdef`
` @note This macro is a no-op if @kconfig{CONFIG_PM_DEVICE} is not enabled.` same for `PM_DEVICE_DT_GET`.
That could simplify it. I know that you have an additional parameter that is passed only this option is enabled,
but you still could have something like:
```
DEVICE_DT_DEFINE(INTEL_ADSP_IPC_HOST_DTNODE, dt_init, PM_DEVICE_DT_GET(INTEL_ADSP_IPC_HOST_DTNODE),
	&ipc_host_data, &ipc_host_config, PRE_KERNEL_2, 0, COND_CODE_1(CONFIG_PM_DEVICE, (&ipc_power_control_apii), (NULL)));
```